clients/notification-kafka/src/main/java/org/eclipse/hono/client/notification/KafkaBasedNotificationReceiver.java

```diff
@@ -0,0 +1,136 @@
+    @SuppressWarnings("rawtypes")
+    private final Map<Class<? extends AbstractNotification>, Handler> handlerPerType = new HashMap<>();
+    private final Set<String> topics = new HashSet<>();
+
+    private Supplier<Consumer<String, Buffer>> kafkaConsumerSupplier;
+    private HonoKafkaConsumer honoKafkaConsumer;
+    private boolean started = false;
+
+    /**
+     * Creates a client for consuming notifications.
+     *
+     * @param vertx The vert.x instance to be used.
+     * @param consumerConfig The configuration for the Kafka consumer.
+     *
+     * @throws NullPointerException if any parameter is {@code null}.
+     * @throws IllegalArgumentException if consumerConfig does not contain a valid configuration (at least a bootstrap
+     *             server).
+     */
+    public KafkaBasedNotificationReceiver(final Vertx vertx, final Map<String, String> consumerConfig) {
+
+        Objects.requireNonNull(vertx);
+        Objects.requireNonNull(consumerConfig);
+
+        if (Strings.isNullOrEmpty(consumerConfig.get(AbstractKafkaConfigProperties.PROPERTY_BOOTSTRAP_SERVERS))) {
+            throw new IllegalArgumentException("No Kafka configuration found!");
+        }
+
+        this.vertx = vertx;
+        this.consumerConfig.putAll(consumerConfig);
+    }
+
+    // visible for testing
+    void setKafkaConsumerFactory(final Supplier<Consumer<String, Buffer>> kafkaConsumerSupplier) {
+        this.kafkaConsumerSupplier = Objects.requireNonNull(kafkaConsumerSupplier);
+    }
+
+    @Override
+    public Future<Void> start() {
+        honoKafkaConsumer = new HonoKafkaConsumer(vertx, topics, getRecordHandler(), consumerConfig);
+        Optional.ofNullable(kafkaConsumerSupplier).ifPresent(honoKafkaConsumer::setKafkaConsumerSupplier);
+        return honoKafkaConsumer
+                .start()
+                .onSuccess(ok -> started = true);
+    }
+
+    @Override
+    public <T extends AbstractNotification> void registerConsumer(final Class<T> notificationType,
+            final Handler<T> consumer) {
+
+        if (started) {
+            throw new IllegalStateException("consumers cannot be added when consumer is already started.");
+        }
+
+        topics.add(NotificationTopicHelper.getTopicName(notificationType));
+        handlerPerType.put(notificationType, consumer);
+    }
+
+    private Handler<KafkaConsumerRecord<String, Buffer>> getRecordHandler() {
+        return record -> handleNotification(Json.decodeValue(record.value(), AbstractNotification.class));
+    }
+
+    // visible for testing
+    @SuppressWarnings("unchecked")
+    void handleNotification(final AbstractNotification abstractNotification) {
+        for (final var entry : handlerPerType.entrySet()) {
+            final Class<? extends AbstractNotification> type = entry.getKey();
+
+            if (type.isInstance(abstractNotification)) {
+                entry.getValue().handle(abstractNotification);
+                break;
+            }
+        }
```

Comment:
can't we simply look up the handler by the notification's type directly, i.e.
```java
final var handler = handlerPerType.get(abstractNotification.getClass());
```

Reply:
Right, thank you. I included this,